src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs

```diff
@@ -142,35 +142,72 @@ private BoundExpression CompareWithConstant(BoundExpression input, BoundExpressi
+        private bool? MatchConstantValue(BoundExpression source, TypeSymbol targetType, bool requiredNullTest)
         {
             // use site diagnostics will already have been reported during binding.
             HashSet<DiagnosticInfo> ignoredDiagnostics = null;
-            switch (_compilation.Conversions.ClassifyBuiltInConversion(sourceType, targetType, ref ignoredDiagnostics).Kind)
-            {
-                case ConversionKind.Boxing:
-                case ConversionKind.ImplicitReference:
-                case ConversionKind.Identity:
-                    return true;
-                default:
-                    return false;
-            }
+            var sourceType = source.Type.IsDynamic() ? _compilation.GetSpecialType(SpecialType.System_Object) : source.Type;
+            var conversionKind = _compilation.Conversions.ClassifyConversionFromType(sourceType, targetType, ref ignoredDiagnostics).Kind;
+            var constantResult = Binder.GetIsOperatorConstantResult(sourceType, targetType, conversionKind, source.ConstantValue, requiredNullTest);
+            return
+                constantResult == ConstantValue.True ? true :
+                constantResult == ConstantValue.False ? false :
+                constantResult == null ? (bool?)null :
+                throw ExceptionUtilities.UnexpectedValue(constantResult);
         }
 
         BoundExpression MakeIsDeclarationPattern(SyntaxNode syntax, BoundExpression loweredInput, BoundExpression loweredTarget, bool requiresNullTest)
         {
             var type = loweredTarget.Type;
             requiresNullTest = requiresNullTest && loweredInput.Type.CanContainNull();
 
+            // If the match is impossible, we simply evaluate the input and yield false.
+            var matchConstantValue = MatchConstantValue(loweredInput, type, false);
+            if (matchConstantValue == false)
+            {
+                return _factory.MakeSequence(loweredInput, _factory.Literal(false));
+            }
+
             // It is possible that the input value is already of the correct type, in which case the pattern
             // is irrefutable, and we can just do the assignment and return true (or perform the null test).
-            if (MatchIsIrrefutable(loweredInput.Type, loweredTarget.Type, requiresNullTest))
+            if (matchConstantValue == true)
             {
-                var convertedInput = _factory.Convert(loweredTarget.Type, loweredInput);
-                var assignment = _factory.AssignmentExpression(loweredTarget, convertedInput);
-                return requiresNullTest
-                    ? _factory.ObjectNotEqual(assignment, _factory.Null(type))
-                    : _factory.MakeSequence(assignment, _factory.Literal(true));
+                requiresNullTest = requiresNullTest && MatchConstantValue(loweredInput, type, true) != true;
+                if (loweredInput.Type.IsNullableType())
```

Comment:
>if (loweredInput.Type.IsNullableType()) [](start = 16, length = 39)

Are all branches, including nested ifs and conditional expressions, covered by tests?